File: btd700/hid.py
"""
Cross-platform Linux HID communication layer for Sennheiser BTD dongles.
Implements native Linux /dev/hidraw communication with zero required external dependencies,
with fallback to python-hidapi if available.
"""

import os
import glob
import time
import fcntl
import select
import struct
import threading
from typing import Optional, List, Dict, Any, Callable

from .constants import (
    SENNHEISER_VID,
    BTD600_PID,
    BTD700_PID,
    KNOWN_DEVICES,
    ReportID,
)

# Linux hidraw ioctl numbers
# _IOC(dir, type, nr, size)
# dir: 0=none, 1=write, 2=read, 3=read|write
# 'H' = 0x48 = 72
def _IOC(dir_flag: int, type_char: str, nr: int, size: int) -> int:
    return (dir_flag << 30) | (size << 16) | (ord(type_char) << 8) | nr

def HIDIOCSFEATURE(length: int) -> int:
    return _IOC(3, 'H', 0x06, length)

def HIDIOCGFEATURE(length: int) -> int:
    return _IOC(3, 'H', 0x07, length)

HIDIOCGRAWINFO = _IOC(2, 'H', 0x03, 8)


class HidrawDevice:
    """Direct interface to a Linux /dev/hidraw* device."""

    def __init__(self, dev_path: str, vid: int, pid: int, name: str = "", report_ids: Optional[List[int]] = None):
        self.dev_path = dev_path
        self.vid = vid
        self.pid = pid
        self.name = name
        self.report_ids = report_ids or []
        self._fd: Optional[int] = None
        self._lock = threading.Lock()
        self._rx_thread: Optional[threading.Thread] = None
        self._running = False
        self._callbacks: List[Callable[[bytes], None]] = []

    def open(self) -> bool:
        if self._fd is not None:
            return True
        try:
            self._fd = os.open(self.dev_path, os.O_RDWR | os.O_NONBLOCK)
            return True
        except PermissionError:
            raise PermissionError(
                f"Permission denied accessing {self.dev_path}. "
                "Ensure your user has access to hidraw devices (e.g. install udev rules) or run with appropriate permissions."
            )
        except OSError as e:
            return False

    def close(self):
        self.stop_reader()
        with self._lock:
            if self._fd is not None:
                try:
                    os.close(self._fd)
                except OSError:
                    pass
                self._fd = None

    def write(self, data: bytes) -> int:
        """Write an output report to the hidraw device."""
        with self._lock:
            if self._fd is None:
                raise IOError(f"Device {self.dev_path} is not open")
            return os.write(self._fd, data)

    def read(self, length: int = 64, timeout_ms: int = 1000) -> Optional[bytes]:
        """Synchronously read an input report."""
        if self._fd is None:
            return None
        r, _, _ = select.select([self._fd], [], [], timeout_ms / 1000.0)
        if r:
            try:
                return os.read(self._fd, length)
            except OSError:
                return None
        return None

    def set_feature(self, data: bytes) -> bool:
        """Send a feature report via ioctl HIDIOCSFEATURE."""
        with self._lock:
            if self._fd is None:
                return False
            try:
                buf = bytearray(data)
                fcntl.ioctl(self._fd, HIDIOCSFEATURE(len(buf)), buf)
                return True
            except OSError:
                return False

    def get_feature(self, report_id: int, length: int = 64) -> Optional[bytes]:
        """Get a feature report via ioctl HIDIOCGFEATURE."""
        with self._lock:
            if self._fd is None:
                return None
            try:
                buf = bytearray([report_id] + [0] * (length - 1))
                fcntl.ioctl(self._fd, HIDIOCGFEATURE(len(buf)), buf)
                return bytes(buf)
            except OSError:
                return None

    def register_callback(self, cb: Callable[[bytes], None]):
        if cb not in self._callbacks:
            self._callbacks.append(cb)

    def unregister_callback(self, cb: Callable[[bytes], None]):
        if cb in self._callbacks:
            self._callbacks.remove(cb)

    def start_reader(self):
        """Start asynchronous background read loop for event-driven updates."""
        if self._running or self._fd is None:
            return
        self._running = True
        self._rx_thread = threading.Thread(target=self._reader_loop, daemon=True, name=f"hidraw-reader-{os.path.basename(self.dev_path)}")
        self._rx_thread.start()

    def stop_reader(self):
        self._running = False
        if self._rx_thread and self._rx_thread.is_alive():
            self._rx_thread.join(timeout=0.5)
        self._rx_thread = None

    def _reader_loop(self):
        while self._running and self._fd is not None:
            try:
                r, _, _ = select.select([self._fd], [], [], 0.1)
                if r:
                    data = os.read(self._fd, 128)
                    if data:
                        for cb in list(self._callbacks):
                            try:
                                cb(data)
                            except Exception:
                                pass
            except (OSError, ValueError):
                break


def find_sennheiser_hid_devices() -> List[Dict[str, Any]]:
    """
    Enumerate connected Sennheiser BTD 600 / BTD 700 devices on Linux.
    Parses /sys/class/hidraw and inspection of report descriptors.
    """
    found = []
    
    # Scan /sys/class/hidraw
    hidraw_dirs = sorted(glob.glob("/sys/class/hidraw/hidraw*"))
    for hdir in hidraw_dirs:
        dev_name = os.path.basename(hdir)
        dev_path = f"/dev/{dev_name}"
        uevent_file = os.path.join(hdir, "device", "uevent")
        if not os.path.exists(uevent_file):
            continue
        
        vid = None
        pid = None
        name = ""
        serial = ""
        
        try:
            with open(uevent_file, "r") as f:
                for line in f:
                    line = line.strip()
                    if line.startswith("HID_ID="):
                        # Format: HID_ID=bustype:vendor:product
                        parts = line.split("=")[1].split(":")
                        if len(parts) >= 3:
                            vid = int(parts[1], 16)
                            pid = int(parts[2], 16)
                    elif line.startswith("HID_NAME="):
                        name = line.split("=")[1].strip('"')
        except OSError:
            continue
        
        if vid == SENNHEISER_VID and pid in (BTD600_PID, BTD700_PID):
            # Parse report descriptor to find supported report IDs
            report_ids = []
            desc_file = os.path.join(hdir, "device", "report_descriptor")
            if os.path.exists(desc_file):
                try:
                    with open(desc_file, "rb") as f:
                        desc = f.read()
                    # Search for 0x85 <report_id> in report descriptor
                    i = 0
                    while i < len(desc) - 1:
                        if desc[i] == 0x85:
                            report_ids.append(desc[i + 1])
                            i += 2
                        else:
                            i += 1
                except OSError:
                    pass
            
            # Read serial number if available
            serial_file = os.path.join(hdir, "device", "..", "serial")
            if not os.path.exists(serial_file):
                # Try higher in sysfs tree
                serial_file = os.path.join(hdir, "device", "..", "..", "serial")
            if os.path.exists(serial_file):
                try:
                    with open(serial_file, "r") as f:
                        serial = f.read().strip()
                except OSError:
                    pass
            
            info = KNOWN_DEVICES.get((vid, pid), {
                "name": "Unknown Sennheiser Device",
                "sku": "",
                "type": None,
                "reports": [],
            })
            
            found.append({
                "dev_path": dev_path,
                "sys_path": hdir,
                "vid": vid,
                "pid": pid,
                "model_name": info.get("name", "Sennheiser Dongle"),
                "product_name": name,
                "serial": serial,
                "report_ids": sorted(list(set(report_ids))),
                "device_info": info,
            })
            
    return found

File: src/hid.rs
//! Linux /dev/hidraw interface for Sennheiser USB dongles.

use std::fs::{self, File, OpenOptions};
use std::io::{self, Write};
use std::os::unix::fs::OpenOptionsExt;
use std::os::unix::io::AsRawFd;
use std::path::Path;
use serde::{Deserialize, Serialize};

use crate::constants::{PID_BTD600, PID_BTD700, VID_SENNHEISER};

const fn ioc(dir: u32, type_char: u8, nr: u32, size: u32) -> u64 {
    ((dir as u64) << 30) | ((size as u64) << 16) | ((type_char as u64) << 8) | (nr as u64)
}

#[inline]
pub fn hidiocsfeature(len: usize) -> u64 {
    ioc(3, b'H', 0x06, len as u32)
}

#[inline]
pub fn hidiocgfeature(len: usize) -> u64 {
    ioc(3, b'H', 0x07, len as u32)
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DeviceInfo {
    pub dev_path: String,
    pub sys_path: String,
    pub vid: u16,
    pub pid: u16,
    pub model_name: String,
    pub product_name: String,
    pub serial: String,
    pub report_ids: Vec<u8>,
}

pub fn find_sennheiser_hid_devices() -> Vec<DeviceInfo> {
    let mut found = Vec::new();
    let sys_hidraw = Path::new("/sys/class/hidraw");
    if !sys_hidraw.exists() {
        return found;
    }

    let entries = match fs::read_dir(sys_hidraw) {
        Ok(e) => e,
        Err(_) => return found,
    };

    for entry in entries.flatten() {
        let hdir = entry.path();
        let dev_name = hdir.file_name().and_then(|n| n.to_str()).unwrap_or("");
        let dev_path = format!("/dev/{dev_name}");

        let uevent_file = hdir.join("device").join("uevent");
        if !uevent_file.exists() {
            continue;
        }

        let mut vid = 0u16;
        let mut pid = 0u16;
        let mut name = String::new();

        if let Ok(content) = fs::read_to_string(&uevent_file) {
            for line in content.lines() {
                if let Some(rest) = line.strip_prefix("HID_ID=") {
                    let parts: Vec<&str> = rest.split(':').collect();
                    if parts.len() >= 3 {
                        vid = u16::from_str_radix(parts[1], 16).unwrap_or(0);
                        pid = u16::from_str_radix(parts[2], 16).unwrap_or(0);
                    }
                } else if let Some(rest) = line.strip_prefix("HID_NAME=") {
                    name = rest.trim_matches('"').to_string();
                }
            }
        }

        if vid == VID_SENNHEISER && (pid == PID_BTD600 || pid == PID_BTD700) {
            let mut report_ids = Vec::new();
            let desc_file = hdir.join("device").join("report_descriptor");
            if let Ok(bytes) = fs::read(&desc_file) {
                let mut i = 0;
                while i < bytes.len().saturating_sub(1) {
                    if bytes[i] == 0x85 {
                        report_ids.push(bytes[i + 1]);
                        i += 2;
                    } else {
                        i += 1;
                    }
                }
            }
            report_ids.sort_unstable();
            report_ids.dedup();

            let mut serial = String::new();
            let candidates = [
                hdir.join("device").join("..").join("serial"),
                hdir.join("device").join("..").join("..").join("serial"),
            ];
            for cand in &candidates {
                if cand.exists() {
                    if let Ok(s) = fs::read_to_string(cand) {
                        serial = s.trim().to_string();
                        break;
                    }
                }
            }

            let model_name = match pid {
                PID_BTD600 => "Sennheiser BTD 600",
                PID_BTD700 => "Sennheiser BTD 700",
                _ => "Sennheiser Dongle",
            }
            .to_string();

            found.push(DeviceInfo {
                dev_path,
                sys_path: hdir.to_string_lossy().to_string(),
                vid,
                pid,
                model_name,
                product_name: name,
                serial,
                report_ids,
            });
        }
    }

    found
}

pub struct HidrawDevice {
    pub info: DeviceInfo,
    file: Option<File>,
}

impl HidrawDevice {
    pub fn new(info: DeviceInfo) -> Self {
        Self { info, file: None }
    }

    pub fn open(&mut self) -> io::Result<()> {
        let file = OpenOptions::new()
            .read(true)
            .write(true)
            .custom_flags(libc::O_NONBLOCK)
            .open(&self.info.dev_path)?;
        self.file = Some(file);
        Ok(())
    }

    pub fn close(&mut self) {
        self.file = None;
    }

    pub fn is_open(&self) -> bool {
        self.file.is_some()
    }

    pub fn write_report(&mut self, data: &[u8]) -> io::Result<usize> {
        let res = {
            let f = self.file.as_mut().ok_or_else(|| {
                io::Error::new(io::ErrorKind::NotConnected, "Device is not open")
            })?;
            f.write(data)
        };

        match res {
            Ok(n) => Ok(n),
            Err(e) => {
                // Fallback to HIDIOCSFEATURE ioctl if standard write fails on some kernels/interfaces
                if self.set_feature(data).is_ok() {
                    Ok(data.len())
                } else {
                    Err(e)
                }
            }
        }
    }

    pub fn read_report(&mut self, timeout_ms: u64) -> io::Result<Option<Vec<u8>>> {
        let f = self.file.as_mut().ok_or_else(|| {
            io::Error::new(io::ErrorKind::NotConnected, "Device is not open")
        })?;
        let fd = f.as_raw_fd();

        let mut pfd = libc::pollfd {
            fd,
            events: libc::POLLIN,
            revents: 0,
        };

        let ret = unsafe { libc::poll(&mut pfd, 1, timeout_ms as libc::c_int) };
        if ret <= 0 {
            return Ok(None);
        }

        if (pfd.revents & libc::POLLIN) != 0 {
            let mut buf = vec![0u8; 128];
            let n = unsafe {
                libc::read(fd, buf.as_mut_ptr() as *mut libc::c_void, buf.len())
            };
            if n > 0 {
                buf.truncate(n as usize);
                return Ok(Some(buf));
            }
        }

        Ok(None)
    }

    pub fn set_feature(&mut self, data: &[u8]) -> io::Result<()> {
        let f = self.file.as_mut().ok_or_else(|| {
            io::Error::new(io::ErrorKind::NotConnected, "Device is not open")
        })?;
        let fd = f.as_raw_fd();
        let cmd = hidiocsfeature(data.len());
        let ret = unsafe {
            libc::ioctl(fd, cmd as libc::c_ulong, data.as_ptr())
        };
        if ret < 0 {
            return Err(io::Error::last_os_error());
        }
        Ok(())
    }

    pub fn get_feature(&mut self, report_id: u8, len: usize) -> io::Result<Vec<u8>> {
        let f = self.file.as_mut().ok_or_else(|| {
            io::Error::new(io::ErrorKind::NotConnected, "Device is not open")
        })?;
        let fd = f.as_raw_fd();
        let mut buf = vec![0u8; len];
        buf[0] = report_id;
        let cmd = hidiocgfeature(len);
        let ret = unsafe {
            libc::ioctl(fd, cmd as libc::c_ulong, buf.as_mut_ptr())
        };
        if ret < 0 {
            return Err(io::Error::last_os_error());
        }
        Ok(buf)
    }
}
